### ISSUE TITLE
rewriter: `static` fns with the same name result in duplicate `extern IA2_fn_ptr_* __ia2_*` declarations

### ISSUE THREAD

**Issue by kkysen** [immunant/IA2-Phase2#428]
When there are multiple `static` functions with the same name, they result in duplicate `extern IA2_fn_ptr_* __ia2_*` declarations in `callgate_wrapper.h`.  I'm not certain this an error, since other errors are in the way first, but it seems like an error.

For example, in `dav1d`, this happens with `init_internal`, which is defined in `lib.c` and `scan.c` and used when calling `pthread_once`.

**Comment by ayrtonm** [comment]
We took the issue of call gates for static functions with the same name into account in a few places but ultimately decided to declare them all in the output header which prevents them. This was just to minimize the number of build artifacts and simplify the build process, but I think there's multiple ways we might be able to handle this.

>  I'm not certain this an error, since other errors are in the way first, but it seems like an error. For example, in dav1d, this happens with init_internal, which is defined in lib.c and scan.c and used when calling pthread_once.

If both these files define a `static` `init_internal` function and take their address (must happen in both .c's) then it's an issue we need to address.

**Comment by ayrtonm** [comment]
`-Wl,-zunique-symbol` might come in handy here if call gate names cannot be local.

**Comment by ayrtonm** [comment]
`-Wl,-zunique-symbol` might come in handy here if call gate names cannot be local.

**Comment by ayrtonm** [comment]
I think the easiest thing to handle the most common cases of this would be to assert that duplicate static functions have the same function signature and generate only one declaration in the header.

**Comment by kkysen** [comment]
> I think the easiest thing to handle the most common cases of this would be to assert that duplicate static functions have the same function signature and generate only one declaration in the header.

Yeah that would improve things.  Move the error to the rewriter, and if the functions are the same, then just remove duplicates.

**Comment by ayrtonm** [comment]
My tests for this cover static functions defined in headers, duplicate definitions of static functions in source files and different definitions of static functions with the same name in source files. @kkysen any other cases I should cover?

**Comment by ayrtonm** [comment]
oh I just realized the duplicate declarations don't actually cause an error and if the function signatures don't match it'll trigger a compile error. Since the generated header isn't considered developer-facing this is much lower priority than I thought. I'll push the tests I wrote and the change I made for #414 (it's needed for static functions defined in headers) but I don't think it's worth spending time on moving the errors to the rewriter for now.

**Comment by kkysen** [comment]
> My tests for this cover static functions defined in headers, duplicate definitions of static functions in source files and different definitions of static functions with the same name and function signature in source files. @kkysen any other cases I should cover?

That sounds like all of them to me.

> oh I just realized the duplicate declarations don't actually cause an error and if the function signatures don't match it'll trigger a compile error. Since the generated header isn't considered developer-facing this is much lower priority than I thought. I'll push the tests I wrote and the change I made for #414 (it's needed for static functions defined in headers) but I don't think it's worth spending time on moving the errors to the rewriter for now.

Is the only solution for now to just `IA2_IGNORE` them?  And be lucky that they don't actually cross compartment boundaries?
